### PR TITLE
auth LDAP + user provisioning API

### PR DESCRIPTION
**Pull request by dnutan**
Changes to be carefully reviewed.

Note: the following **Directory Settings** options are not described in the documentation:
- Dynamic Group Member URL
- Nested Groups
- Paging chunksize

### CODE REVIEW

**Comment by MorrisJobke** [comment]
cc @blizzz

File: admin_manual/configuration_user/user_auth_ldap.rst [hunk 1]
```diff
@@ -161,21 +160,21 @@ LDAP filter.
 LDAP Username:
   If this value is checked, the login value will be compared to the username in
   the LDAP directory. The corresponding attribute, usually *uid* or
-  *samaccountname* will be detected automatically by Nextcloud.
+  *samaccountname* will be detected automatically by Nextcloud.
```
Comment (blizzz) [inline admin_manual/configuration_user/user_auth_ldap.rst]: trailing white space

File: admin_manual/configuration_user/user_auth_ldap.rst [hunk 2]
```diff
@@ -161,21 +160,21 @@ LDAP filter.
...
   If this value is checked, the login value will be compared to the username in
   the LDAP directory. The corresponding attribute, usually *uid* or
-  *samaccountname* will be detected automatically by Nextcloud.
+  *samaccountname* will be detected automatically by Nextcloud. 
 
 LDAP Email Address:
   If this value is checked, the login value will be compared to an email address
   in the LDAP directory; specifically, the *mailPrimaryAddress* and *mail* 
-  attributes.
+  attributes.
```
Comment (blizzz) [inline admin_manual/configuration_user/user_auth_ldap.rst]: trailing white space

File: admin_manual/configuration_user/user_auth_ldap.rst [hunk 3]
```diff
@@ -161,21 +160,21 @@ LDAP filter.
...
   If this value is checked, the login value will be compared to an email address
   in the LDAP directory; specifically, the *mailPrimaryAddress* and *mail* 
-  attributes.
+  attributes. 
 
 Other Attributes:
   This multi-select box allows you to select other attributes for the 
   comparison. The list is generated automatically from the user object 
-  attributes in your LDAP server.
+  attributes in your LDAP server.
```
Comment (blizzz) [inline admin_manual/configuration_user/user_auth_ldap.rst]: trailing white space

File: admin_manual/configuration_user/user_auth_ldap.rst [hunk 4]
```diff
@@ -161,21 +160,21 @@ LDAP filter.
...
   This multi-select box allows you to select other attributes for the 
   comparison. The list is generated automatically from the user object 
-  attributes in your LDAP server.
+  attributes in your LDAP server. 
 
-Edit raw filter instead:
+Edit LDAP Query:
   Clicking on this text toggles the filter mode and you can enter the raw LDAP 
-  filter directly.
+  filter directly.
```
Comment (blizzz) [inline admin_manual/configuration_user/user_auth_ldap.rst]: trailing white space

File: admin_manual/configuration_user/user_auth_ldap.rst [hunk 5]
```diff
@@ -192,30 +191,30 @@ Edit raw filter instead:
...
 .. figure:: ../images/ldap-wizard-4-group.png
    :alt: Group filter
 
-only those object classes:
+Only these object classes:
   Nextcloud will determine the object classes that are typically available for
   group objects in your LDAP server. Nextcloud will only list object
   classes that return at least one group object. You can select multiple
-  object classes. A typical object class is "group", or "posixGroup".
+  object classes. A typical object class is "group", or "posixGroup".
```
Comment (blizzz) [inline admin_manual/configuration_user/user_auth_ldap.rst]: trailing white space

File: admin_manual/configuration_user/user_auth_ldap.rst [hunk 6]
```diff
@@ -192,30 +191,30 @@ Edit raw filter instead:
...
-  object classes. A typical object class is "group", or "posixGroup".
+  object classes. A typical object class is "group", or "posixGroup". 
 
-only from those groups:
+Only from these groups:
   Nextcloud will generate a list of available groups found in your LDAP server. 
-  and then you select the group or groups that get access to your Nextcloud 
-  server.
+  Then you select the group or groups that get access to your Nextcloud 
+  server.
```
Comment (blizzz) [inline admin_manual/configuration_user/user_auth_ldap.rst]: trailing white space

File: admin_manual/configuration_user/user_auth_ldap.rst [hunk 7]
```diff
@@ -192,30 +191,30 @@ Edit raw filter instead:
...
-  and then you select the group or groups that get access to your Nextcloud 
-  server.
+  Then you select the group or groups that get access to your Nextcloud 
+  server. 
 
-Edit raw filter instead:
+Edit LDAP Query:
   Clicking on this text toggles the filter mode and you can enter the raw LDAP 
-  filter directly.
+  filter directly.
```
Comment (blizzz) [inline admin_manual/configuration_user/user_auth_ldap.rst]: trailing white space

File: admin_manual/configuration_user/user_auth_ldap.rst [hunk 8]
```diff
@@ -250,38 +249,38 @@ Connection Settings
 Configuration Active:
   Enables or Disables the current configuration. By default, it is turned off. 
   When Nextcloud makes a successful test connection it is automatically turned 
-  on.
+  on.
```
Comment (blizzz) [inline admin_manual/configuration_user/user_auth_ldap.rst]: trailing white space

File: admin_manual/configuration_user/user_auth_ldap.rst [hunk 9]
```diff
@@ -250,38 +249,38 @@ Connection Settings
...
   When Nextcloud makes a successful test connection it is automatically turned 
-  on.
+  on. 
 
 Backup (Replica) Host:
   If you have a backup LDAP server, enter the connection settings here. 
   Nextcloud will then automatically connect to the backup when the main server 
   cannot be reached. The backup server must be a replica of the main server so 
-  that the object UUIDs match.
+  that the object UUIDs match.
```
Comment (blizzz) [inline admin_manual/configuration_user/user_auth_ldap.rst]: trailing white space

File: admin_manual/configuration_user/user_auth_ldap.rst [hunk 10]
```diff
@@ -250,38 +249,38 @@ Connection Settings
...
+  that the object UUIDs match. 
 
   Example:
 
   * *directory2.my-company.com*
 
 Backup (Replica) Port:
   The connection port of the backup LDAP server. If no port is given,
-  but only a host, then the main port (as specified above) will be used.
+  but only a host, then the main port (as specified above) will be used.
```
Comment (blizzz) [inline admin_manual/configuration_user/user_auth_ldap.rst]: trailing white space

File: admin_manual/configuration_user/user_auth_ldap.rst [hunk 11]
```diff
@@ -250,38 +249,38 @@ Connection Settings
...
+  but only a host, then the main port (as specified above) will be used. 
 
   Example:
 
   * *389*
 
 Disable Main Server:
   You can manually override the main server and make Nextcloud only connect to
-  the backup server. This is useful for planned downtimes.
+  the backup server. This is useful for planned downtimes.
```
Comment (blizzz) [inline admin_manual/configuration_user/user_auth_ldap.rst]: trailing white space

File: admin_manual/configuration_user/user_auth_ldap.rst [hunk 12]
```diff
@@ -250,38 +249,38 @@ Connection Settings
...
   * *389*
 
 Disable Main Server:
   You can manually override the main server and make Nextcloud only connect to
-  the backup server. This is useful for planned downtimes.
+  the backup server. This is useful for planned downtimes. 
 
 Turn off SSL certificate validation:
-  Turns off SSL certificate checking. Use it for testing only!
+  Turns off SSL certificate checking. Use it for testing only!
```
Comment (blizzz) [inline admin_manual/configuration_user/user_auth_ldap.rst]: trailing white space

File: admin_manual/configuration_user/user_auth_ldap.rst [hunk 13]
```diff
@@ -250,38 +249,38 @@ Connection Settings
...
 Turn off SSL certificate validation:
-  Turns off SSL certificate checking. Use it for testing only!
+  Turns off SSL certificate checking. Use it for testing only! 
 
 Cache Time-To-Live:
   A cache is introduced to avoid unnecessary LDAP traffic, for example caching 
   usernames so they don't have to be looked up for every page, and speeding up 
   loading of the Users page. Saving the configuration empties the cache. The 
-  time is given in seconds.
+  time is given in seconds.
```
Comment (blizzz) [inline admin_manual/configuration_user/user_auth_ldap.rst]: trailing white space

File: admin_manual/configuration_user/user_auth_ldap.rst [hunk 14]
```diff
@@ -311,7 +310,7 @@ User Display Name Field:
 2nd User Display Name Field:  
   An optional second attribute displayed in brackets after the display name, 
   for example using the ``mail`` attribute displays as ``Molly Foo 
-  (molly@example.com)``.
+  (molly@example.com)``.
```
Comment (blizzz) [inline admin_manual/configuration_user/user_auth_ldap.rst]: trailing white space

**Pull request merged**